### ISSUE TITLE
HTML input elements with type number should only be valid when a number is passed.

### ISSUE THREAD

**Issue by rbi** [HtmlUnit/htmlunit#321]
Currently the `checkValidity()` method on an input element of type number only checks the lower and the upper bound of the value that is passed. When something is passed that isn't a number at all `checkValidity()` returns `true` although it should return `false`.

I'll open a pull request that demonstrates the problem and suggests a fix in a minute.

**Comment by rbri** [comment]
Sorry but your test was not passing with real browsers. Please have a look at the WebDriverTestCase class javadoc for more.

Have written some more tests based on your suggestion/problem and updated the implementation. If the CI passes i will make a new SNAPSHOT build available. You can have a look at twitter to get informed about new snapshot builds.

Will be great if you can test test new snapshot (if available) and provide feedback.

Many thanks for you contribution.

**Comment by rbi** [comment]
Thanks for looking into it. I must admit that I didn't check the outcome in real browsers for the input values I used in the test.

I can checkout a SNAPSHOT build on Wednesday with the test code I've run that issue into.

**Comment by rbri** [comment]
Great thanks.

**Comment by rbi** [comment]
I've tested the SNAPSHOT with my real test case. Unfortunately it doesn't work. I use Selenium. This is my Test case:

~~~
    @Test
    void test() {
        WebDriver driver = createDriverAndNavigateToPage();
        
        WebElement numberInputField = driver.findElement(By.id("someNumberInputField"));
        
        numberInputField.clear();
        numberInputField.sendKeys("abc");

        boolean isValid = (boolean) ((JavascriptExecutor) driver).executeScript("return arguments[0].checkValidity();", numberInputField);
        assertFalse(isValid);
    }
~~~

When I create a `HtmlUnitDriver` in `createDriverAndNavigateToPage()` the test fails. When I create a `FirefoxDriver` it works.

I've created a break point in `HtmlNumberInput#isValid`. At this code:

~~~
final String valueAttr = getValueAttribute();
        if (!valueAttr.isEmpty()) {
~~~

`valueAttr` is empty so all tests are skipped and `true` is returned.

What I noticed with the `FirefoxDriver` is that the `value` of the `<input type="number">` is an empty String when garbage is input although visually the garbage is displayed in Firefox. In HtmlUnit the value is now an empty String too. I think it wasn't empty before though I'm not sure. Despite the empty value the `checkValidity()` Method in Firefox returns false while in HtmlUnit true is returned.

I' used the following code to get the value of the input Element

~~~
String value = (String) ((JavascriptExecutor) driver).executeScript("return arguments[0].value;", numberInputField);
~~~

**Comment by rbri** [comment]
Ok, will have a look again. Thanks for testing.

You are sure you use the latest snapshot (because the WebDriver has the latest release as dependency)?

**Comment by rbi** [comment]
I think it should be the right Snapshot:
* My eclipse automagically downloaded the source code JAR with the changes you've done in https://github.com/HtmlUnit/htmlunit/commit/0841a590d252bda5c6f94f1ff7c7256aa33a0f19
* I could make a break point at `if (!valueAttr.isEmpty()) {` which was introduced by that commit.
* I didn't use `2.48.0-SNAPSHOT` before so this morning should be the first time when it was downloaded.

**Comment by rbri** [comment]
Can you please verify with the latest snapshot build. I tried to come closer to real browsers.

**Comment by rbi** [comment]
Still not working. Maybe we expect a different outputs from the `checkValidity()` Method. For example this test case:

`HtmlNumberInputTest#defaultValuesInvalidValue()`

https://github.com/HtmlUnit/htmlunit/blob/2680c342ef12028db398ac3c5b5dc14a91a59fb8/src/test/java/com/gargoylesoftware/htmlunit/html/HtmlNumberInputTest.java#L641

You put in "abc" into an input element of type "number" and expect `checkValidity()` to return `true`. I expect it to return `false`. With the Chrome and Firefox Selenium Drivers I get false in that cases. At least when I explicitly do `sendKeys("abc")`.

**Comment by rbri** [comment]
OK - a bit clarification about HtmlUnit Tests :-)

The test class extends WebDriverTestCase - this means i have a config and i can run the unchanged class once using Selenium/WebDriver and once with HtmlUnit. Doing this offers the possibility to compare the real browser with HtmlUnit without doing it manually. (You can have a look at the javadoc for more details about this)

All tests from this class are passing here when using the real browser. But in some cases HtmlUnit  still gives different results - see the @HtmlUnitNYI annotation.

This also means that the test method you are talking about does not call sendKeys. But there are other methods testing this sendKeys cases.

Maybe there is a case missing - will be great if you can provide your case - at best in the form of something comparable to e.g. typeInvalidChars().

**Comment by rbri** [comment]
Or maybe simpler - can you provide the definition of your numberInputField.

**Comment by rbi** [comment]
This is my input element.

```
<input id="autoSignoutElem" type="number" min="0" max="999" value="0" name="logoutTimeInMinutes" data-toggle="popover" title="Numbers between 1 and 999 are allowed only">
```

and this is the form element around it. I just changed the action from the original.

```
<form action="/censored/path#_censoredaction" method="post" id="viewer_signout_form">
```

For writing a test case I need some time.

**Comment by rbri** [comment]
Thanks - will have a look at my tests again in the evening.

**Comment by rbi** [comment]
Or maybe a step back: I have a requirement like "The auto sign out time must be a value between 0 and 999". As tester I want to test when I enter non-numbers it should not work. Entering non-numbers is `sendKeys()`. How would I ideally check that this does not work?

**Comment by rbri** [comment]
Good point (for the long answer have a look at www.wetator.org or https://www.informatik-aktuell.de/entwicklung/methoden/test-von-webanwendungen-auf-das-werkzeug-kommt-es-an.html).

So from my point of view you should use sendKeys() and then click on the submit button and check for the error msg.

**Comment by rbri** [comment]
In between i have added another testcase and also did a minor fix for the HtmlUnit driver but i fear this will not solve your problem.

**Comment by rbri** [comment]
Or maybe you can check for the error msg if the :invalid pseudo class is used to show some error. In this case i have to implement the support for this pseudo class in HtmlUnit :-)

**Comment by rbi** [comment]
I guess for the :invalid pseudo class error message the number input field still needs to be detected as invalid. I suppose the error message and the `checkValid()` method share the same validation logic or do they?

**Comment by rbri** [comment]
Yes of course. While reading the whole thread again i will add another testcase later today. Maybe there is hope to fix more....

**Comment by rbri** [comment]
Hope i finally got your point:

    /**
     * @throws Exception if the test fails
     */
    @Test
    @Alerts(DEFAULT = "-0-0-true",
            FF = "-0-0-false",
            FF78 = "-0-0-false")
    @HtmlUnitNYI(IE = "-0-0-false")
    public void issue321() throws Exception {
        final String html = "<html>\n"
                + "<head>\n"
                + "<script>\n"
                + LOG_TITLE_FUNCTION
                + "  function test() {\n"
                + "    var input = document.getElementById('inpt');\n"
                + "    log(input.value + '-' "
                                + "+ input.defaultValue + '-' "
                                + "+ input.getAttribute('value')+ '-' "
                                + "+ input.checkValidity());\n"
                + "  }\n"
                + "</script>\n"
                + "</head>\n"
                + "<body>\n"
                + "  <input type='number' id='inpt' min='0' max='999' value='0' />\n"
                + "  <button id='check' onclick='test()');'>"
                + "DoIt</button>\n"
                + "</body>\n"
                + "</html>";
        final WebDriver driver = loadPage2(html);

        final WebElement input = driver.findElement(By.id("inpt"));

        input.clear();
        input.sendKeys("abc");
        ((JavascriptExecutor) driver).executeScript("test();");
        verifyTitle2(driver, getExpectedAlerts());
    }

This test passes with Selenium and with HtmlUnit. I also have tested this manually - in ff you can input characters and the entry is invalid afterwards. Chrome does not allow chars - your keystrokes are ignored.

So i think HtmlUnit is now correct. Regarding your testing you can only prove the application behavior.
* (try) to type some chars into the number field
* click submit
** this will fail in ff (like in real browsers)
** this will pass in chrome and set the timeout to 0 (like in real chrome)

If your browser behaves differently maybe some js is involved also. Then we have do dig again.

Hope that help...

**Comment by rbi** [comment]
Thank you for your great support. Unfortunately I have to wait until Wednesday again to test it against my real world code. The test case in your last comment looks indeed like the problem I have.

**Comment by rbri** [comment]
no problem, there are still some other tasks to work on :-)

**Comment by rbi** [comment]
Well still not working. I type in "abc" in an input of type "number" and `checkValidity()` returns true.

I haven't found the `public void issue321()` test case you pasted two comments ago in the code. I therefore pasted it into the `HtmlNumberInputTest`. It runs green there.

I haven't checked how this test-in-differnt-Browsers-magic works there but you have:

~~~
@Alerts(DEFAULT = "-0-0-true",
        FF = "-0-0-false",
        FF78 = "-0-0-false")
@HtmlUnitNYI(IE = "-0-0-false")
~~~

Aren't you explicitly saying there "OK in real browsers checkValidity() returns false in this case but in HTMLUnit I want it to return true".

*Edit*: I've tried with htmlunit-2.49.0-SNAPSHOT. I've deleted the htmlunit folder in my local Maven Repo and checked that it got recreated and populated with htmlunit v2.49.0-SNAPSHOT

**Comment by rbri** [comment]
Have tied to reach you via private mail - but no response so far. I think we have to talk about this.

**Comment by rbi** [comment]
I wrote you an E-Mail on Saturday. Did you receive it?

**Comment by rbri** [comment]
Mis..... - was moved to spam.
